verified_boot_app_spec.md

```diff
@@ -0,0 +1,634 @@
+
+In the next phase the application analyzes boot options, prompts the user
+whether to trust they key used to sign the image in the boot option.
+
+Once trusted key database is configured, the application creates ephemeral
+Platform Key `PK`, discards its private part, and enrolls the public part into
+`PK` variable to activate UEFI Secure Boot.
+
+Only after the UEFI Secure Boot environment is provisioned and a Platform Key
+is set does the application default to the interactive menu interface on
+subsequent invocations. This ensures a streamlined initial experience focused
+on system provisioning, while providing full transparency and control
+afterward.
+
+The interactive menu interface is designed to provide detailed view on current
+system environment:
+
+- Current boot options and their:
+  - Verification status
+  - Key fingerprint if the image is signed
+  - Key trust status
+- Current trusted key database
+
+As well as options for Soverign Boot reprovisioning.
+
+## 3. Operational Phases
+
+### 3.1 Environment Initialization
+
+The application can only run before `EFI_BOOT_SERVICES.ExitBootServices()` has
+been called.
+
+The application locates all necessary protocols for its operation:
+
+- HOB list (for `EFI_BOOT_MODE` information)
+- Boot Services (for `LoadImage` and `StartImage` and possibly other API)
+- Runtime Services (for Variable services and possibly other API)
+- crypto libraries (for parsing the certificates and signature verification in
+  the images)
+
+The application also consults the `Boot####` and `BootOrder` variables to
+determine the available boot options and priority of bootloader processing.
+
+If the application is run after initial provisioning, it will read out current
+`PK` and `db` and parse its content.
+
+### 3.3 Bootloader Identification and Verification
+
+The application will parse the images of the bootloaders to obtain information
+about the signatures. The application will additionally validate bootloaders
+using cyrptographic libraries. Bootloaders that pass signature checks using
+are listed for potential trust decisions. For images with unknown but
+verifiable keys, the application initiates a trust prompt.
+
+The priority of bootloader processing is determined by `BootOrder` variable
+prepared by the firmware before launching the application.
+
+The information about the bootloader should include:
+
+* The EFI image device path
+* Status of image verification
+* Associated certificates/keys fingerprints
+
+### 3.4 Trust Decision and Key Enrollment
+
+The application parses the certificates embedded inside the PE/COFF image
+format as described in the UEFI specification. For each newly encountered
+signing key fingerprint/certificate discovered during bootloader verification,
+the application prompts the user whether to trust the key/certificate. The
+prompt must include the key/certificate fingerprint. The prompt may display
+additional information:
```

Comment:
"newly encountered" - this implies that if the user chooses to distrust a key, the info about that key and the fact that it is distrusted will be stored somewhere? That sounds useful to keep a user from being re-prompted to trust a key they chose to not trust multiple times, but should be spelled out if that's what we're doing.

Reply:
Yes, the information about distrusted keys is stored as a UEFI variable. Then if the wizard is launched again and user would like to modify policies, they may see that they previously distrusted the key and can change their decision.

The idea was to avoid multiple prompts for the same key over and over.

Reply:
Assuming the user can change their mind later if we're using the DBX, I'm satisfied with this.

Reply:
>  The user can change their mind even if using the DBX, right?

Yes, they can, however that may not necessarily be a good idea.

Given that the wizard will not touch DBX (because it may contain important security mitigations for buggy or insecure images or keys), this database will be shared with what UEFI forum provides as default DBX. Fortunately, the EDK2 sees the default DBX as a single signature list containing all forbidden hashes, and addign new forbidden hashes creates nex tsignature list. The wizard may simply skip the first signature list assuming it is the golden one from UEFI forum, and just look at all other signature lists assuming they were added by the user.

I think that should help with maintaining the DBX properly if we are going to use it.